### ISSUE TITLE
Compiler required by conda environment isn't used?

### ISSUE THREAD

**Issue by angus-g** [OceanParcels/parcels#811]
The file which defines the dependencies in the conda environment, e.g. for Linux [`environment_py3_linux.yml`](https://github.com/OceanParcels/parcels/blob/master/environment_py3_linux.yml) file has `gcc_linux-64` as a dependency. On a fresh install of Ubuntu Eoan without `gcc`, this dependency doesn't actually give `gcc` as an executable on the path. Instead, when the conda environment is activated, the compiler is provided through the `$CC` environment variable:
```sh
$ echo $CC
/home/ubuntu/miniconda3/envs/parcels/bin/x86_64-conda_cos6-linux-gnu-cc
```

There's some provision for this in parcels: the base `Compiler` class will load the compiler from the `$CC` environment variable. But, using this class directly won't work because it doesn't set up the other required compiler flags. This happens in `GNUCompiler`, however this initialises the superclass with an explicit `compiler="gcc"`, so the above logic doesn't apply.

So, under Linux the `gcc_linux-64` dependency isn't (and can't be) used. It looks like the Mac dependencies request clang, which also wouldn't be found by name.

**Comment by erikvansebille** [comment]
Thanks for reporting this @angus-g! Do you have a suggestion for how we could then improve/fix the environment files and/or compiler.py?

**Comment by angus-g** [comment]
It kind of makes sense for the conda environment to be a full turnkey thing that sets everything up, for places like my Ubuntu container without gcc.

A project with a similar just-in-time compilation philosophy is PyOP2, see [`compilation.py`](https://github.com/OP2/PyOP2/blob/master/pyop2/compilation.py). This might make for a good reference for really fleshing things out. A simple solution might be to let the environment's `$CC` override the compiler chosen by `GNUCompiler` (maybe accompanied by a warning):

```python
self._cc = os.environ.get(ccenv, cc)
```

This used to be the case, changed by https://github.com/OceanParcels/parcels/commit/554b19edba88935a3b9c6e9fe5f0583489fe8b80 that I can see, so I wouldn't want to barge straight back to what was there before. Perhaps especially because I can't test the motivating issue #332 without access to a Mac system.

Otherwise, I guess the dependencies could be pared down to not rely on gcc, but you'd probably want to be explicit that having a compiler called `gcc` is a requirement in that case.

**Comment by CKehl** [comment]
Quick thing on it already cause I was working recently with that: as the first thing of any Ubuntu installation, I install gcc, so I may have not encountered that problem. The dependency is specifically there for clusters. On some other UNIX distros, when pulling 'some' gcc via conda, it will give you the gcc toolchain (gcc, libglibc, ...) - which gives a deprecated version that leads to errors within `pandas`. In order to circumvent this, `pandas` maintainers (as many other conda-forge packages too) recommend to use the 'gcc_linux-64' toolchain (gcc_linux-64, libglibc_linux-64). That is the reason why it is in the `environment.yml` dependencies.

I did not go further on the reading with `Compiler` and `GNUCompiler`, as I am quite entangled right now. Generally though, I agree: the correct behaviour is always to (1) detect the platform, to check whether the CC variable is available (UNIX, MacOS) or if it's Microsoft; (2) if GNU/UNIX-alike environment, read the CC variable and use the path given there, unless CC not specified; (3) if CC is not set, apply other detection methods. I am sure there's also a wikipedia entry on the search method hierarchy for the compilers one should be using. Thanks for the hint. We'll have a look into it the next days.

**Comment by CKehl** [comment]
update: I just tried to run py.test and now, I see the trouble - it takes the wrong compiler, indeed.

**Comment by CKehl** [comment]
Hey Angus, I had a look on the example you posted - I thing the important function here is `def sniff_compiler_version(cc)`. This is indeed much more elegant than what we do now.

Due to internal project involvement, I currently do not have the capacity to implement that change myself. You already had a good look into parcels, so feel free to implement that change as a PR. I'd be happy then to test it on the various cluster capabilities we have here to verify the correct functioning. Otherwise we may have the time later in summer to apply the proposed patch.

**Comment by erikvansebille** [comment]
I've just finished a refactoring of the `Compiler` class, based on code developed by @CKehl. See #914. 

@angus-g, could you check if this refactoring fixes the Issue you reported here? Any feedback on how we can further improve?

**Comment by angus-g** [comment]
Thanks Erik, the refactor correctly pulls in a compiler from the `CC` environment variable, so it works for the issue described here! It does make the assumption that if you're specifying that override, you're giving a GCC-compatible compiler -- maybe there could be issues if somebody has a different breed of compiler sitting in `CC` and Parcels has for the moment been ignoring that?

**Comment by erikvansebille** [comment]
Fixed in #914
